checks/tasks/shared.py

```diff
@@ -110,6 +146,36 @@ def do_resolve_a_aaaa(self, qname, *args, **kwargs):
     return af_ip_pairs
 
 
+def do_resolve_mx(self, url, *args, **kwargs):
+    """Resolve the domain's mailservers
+    returns [(mailserver, af_ip_pairs)]
+    """
+    mx_ips_pairs = []
+
+    for rr, _, status in do_mail_get_servers(self, url, *args, **kwargs):
+        if status is not MxStatus.has_mx:
+            continue
+
+        ips = do_resolve_a_aaaa(self, rr)
+        mx_ips_pairs.append((rr, ips))
+
+    return mx_ips_pairs
+
+
+def do_resolve_ns(self, url, *args, **kwargs):
+    """Resolve the domain's nameservers
+    Returns [(nameserver, af_ip_pairs)]
+    """
+    rrset = self.resolve(url, unbound.RR_TYPE_NS)
+    next_label = url
+    while not rrset and "." in next_label:
+        rrset = self.resolve(next_label, unbound.RR_TYPE_NS)
+        next_label = next_label[next_label.find(".")+1:]
+
+    for rr in rrset:
+        yield (rr, do_resolve_a_aaaa(self, rr))
```

Comment:
Personally I find yield / iterators an improvement in readability compared to the 'define a list and append' pattern (e.g. in `do_resolve_mx`). Any downsides I'm missing @gthess?

Reply:
Just readability and expectation since this is a generator function now; it doesn't return a list as the other similar resolve_ funtions. That's fine though, I thought you were trying to do something fancy with async code.